Feature/calendar recurrence rules uwp

ScottBTR opened the pull request:
Added Recurrence rules update & create for UWP. 
- Collection View still not properly reflecting Observable collection at this stage.
- The model of the recurrence rule may want to be altered if we are using UWP as a guide.

ScottBTR committed:
Allow Recurrence Rules to be Updated for UWP (plus some other required changes)

Some redesign might be required for the actual model. But haven't fully fleshed out what can/can't be done cross-platform at this stage.

All i know is UWP doesn't seem to use as many 'lists' as android/ios. e.g. you can't say occurs on the first AND third of each month.

ScottBTR committed:
Recognition of month week+day offset for UWP

ScottBTR committed:
UWP: Create and Update working to a degree, collection view still requires some work.

ScottBTR committed:
UWP+Android+Ios - Can allow for an optional end date, so I've added support for that.

ScottBTR committed:
Using new UWP recurrence rule model

(with a readonly version existing to either allow us to view rules we do not process the power to create, or to revert back to at a later stage)

the iOS and Android is commented out at this stage as they will most likely require reworks once we have a clear vision going forward.

the selection of days of the week works for selection but not initilization will be looking to convert to a list of checkboxes tomorrow if I am unable to solve it within the first hour.

ScottBTR committed:
Converted CollectionView of Observable collection to use Switches

Altered class name TestDumbReference --> DayOfTheWeekSwitch to match UI elements/make logical sense

ScottBTR committed:
Added crossplatform URI field

ScottBTR committed:
Code Tidy up

ScottBTR committed:
UI more Mobile Friendly

ScottBTR committed:
Include the Appointment Organizer in list of Attendees

ScottBTR committed:
Changes to be more consistent with our OSs

- Allow Intervals to be set up to 99 instead of limiting to a drop down up to 10
- Allow an an end point to be set by number of iterations rather than relying on end date.

Improve UI for the new fields/changes

ScottBTR committed:
Updating to UI and initial set up for iOS

Starting work on iOS to be able to read/update recurrence rules following the UI in place

Altering end point UI to match flutters drop down menu control system as well as adding validation that makes sure values exist when looking for a date/occurrences before allowing creation/update of the event to take place

ScottBTR committed:
Updates to UI and cross-platform attendee types implemented

ScottBTR committed:
Android fixes, plus timezone fixes for android + iOS

nickrandolph commented on Samples/Samples/Converters/AttendeeRequiredColorConverter.cs:
```diff
@@ -10,15 +11,19 @@ public class AttendeeRequiredColorConverter : IValueConverter
             }
 
-            if (required)
+            if (attendeeType == AttendeeType.Required)
```
Perhaps use a switch statement here

ScottBTR replied:
Altered in latest commit

nickrandolph commented on Samples/Samples/Converters/EndTypeToBoolConverter.cs:
```diff
@@ -0,0 +1,36 @@
+            }
+            switch (type)
+            {
+                case RecurrenceEndType.Indefinitely:
```
group cases that have the same return value

ScottBTR replied:
Altered in latest commit

nickrandolph commented on Samples/Samples/Converters/EndTypeToBoolConverter.cs:
```diff
@@ -0,0 +1,36 @@
+{
+    public class EndTypeToBoolConverter : IValueConverter
+    {
+        public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
```
Move System.Globalization namespace into using statement

ScottBTR replied:
Done

nickrandolph commented on Samples/Samples/Converters/EndTypeToBoolConverter.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Samples.Converters
+{
+    public class EndTypeToBoolConverter : IValueConverter
```
Change class name to RecurrenceEndTypeToBoolConverter
I would also suggest it might be better to encapsulate this functionality in the viewmodel (ie expose a different property) as it's hard to understand what this converter is used for

ScottBTR replied:
Have altered the class as suggested, but what would you suggest for the exposing a different property?

nickrandolph commented on Samples/Samples/Converters/ListSelectorArgsConverter.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Samples.Converters
+{
+    public class ListSelectorArgsConverter : IValueConverter
```
Is this converter necessary?

ScottBTR replied:
Not anymore, have removed as suggested

nickrandolph commented on Samples/Samples/Converters/UntilTypeToBoolConverter.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Samples.Converters
+{
+    public class UntilTypeToBoolConverter : IValueConverter
```
See comments on EndTypeToBoolConverter

ScottBTR replied:
Altered in latest commit

nickrandolph commented on Samples/Samples/View/CalendarEventAddPage.xaml:
```diff
@@ -61,11 +66,20 @@
+               Text="{Binding Url}"
+               Grid.Row="4"
+               Grid.Column="1"/>
+
```
Remove spare line

ScottBTR replied:
Altered in latest commit

nickrandolph commented on Samples/Samples/View/CalendarEventAddPage.xaml.cs:
```diff
@@ -11,5 +11,37 @@ public CalendarEventAddPage()
+        int maxIntervalLength = 2;
+
+        void Entry_OnTextChanged(object sender, TextChangedEventArgs e)
+        {
```
What rule are we trying to enforce here? This type of functionality should be in the view model

ScottBTR replied:
Shouldn't really be required, realised I could just use MaxLength="2" combined with Keyboard="Number" to prevent more than 2 digits being entered

nickrandolph commented on Samples/Samples/View/CalendarEventPage.xaml.cs:
```diff
@@ -104,7 +104,7 @@ async void OnEditEventButtonClicked(object sender, EventArgs e)
             var modal = new CalendarEventAddPage();
 
-            var calendarName = (await Calendar.GetCalendarsAsync()).Where(x => x.Id == ViewModel.CalendarId).First().Name;
+            var calendarName = (await Calendar.GetCalendarsAsync()).First(x => x.Id == ViewModel.CalendarId).Name;
```
Use FirstOrDefault and check for null to prevent exceptions when calendarid not found

ScottBTR replied:
altered in latest commit

nickrandolph commented on Samples/Samples/ViewModel/CalendarEventAddAttendeeViewModel.cs:
```diff
@@ -49,12 +49,19 @@ public string EmailAddress
         }
 
-        bool required;
+        public ObservableCollection<string> AttendeeTypes { get; } = new ObservableCollection<string>()
```
This doesn't need to be an observablecollection since it's a static list

ScottBTR replied:
altered in latest commit

nickrandolph commented on Samples/Samples/ViewModel/CalendarEventAddViewModel.cs:
```diff
@@ -152,6 +236,267 @@ public TimeSpan EndTime
+                    OnPropertyChanged(nameof(IsMonthlyOrYearly));
+                    OnPropertyChanged(nameof(IsYearly));
+                    OnPropertyChanged(nameof(SelectedRecurrenceTypeDisplay));
+                    RecurrenceInterval = 1;
```
Why is this being set here? If you're populating an existing recurrence pattern you probably don't want this set

ScottBTR replied:
Should be getting handled properly behind the scenes now, it was a fix to some buggy behaviour that was occurring previously

nickrandolph commented on Samples/Samples/ViewModel/CalendarEventAddViewModel.cs:
```diff
@@ -152,6 +236,267 @@ public TimeSpan EndTime
+            RecurrenceFrequency.Yearly.ToString()
+        };
+
+        string selectedRecurrenceType = RecurrenceFrequency.None.ToString();
```
Wouldn't it be better to have this as a RecurrencyFrequency rather than a string?

ScottBTR replied:
Adjusted in latest commit, as well as fixed all other related parameters that were using strings instead of well formed enums.

ScottBTR committed:
Updates

- Recurrence Rules moved to converter for android/ios to work
- Start date now consistent in design when all day event
- Testing out ToLocalTimeConverter to help with date consistency issues, currently only partially working with android
- Getting event INSTANCES in trial with android
- Added url validator
- In process of adding attendee email validator
- In process of deletion by event instance instead of only by entire instance
- UWP can kinda delete bookings now, but has to be done through windows UI
- In process of adding attendees for iOS

ScottBTR committed:
Get TimeZones && Recurrence Rule Frequencies working more reliably accross all platforms.

ScottBTR committed:
Meeting PR Feedback

ScottBTR committed:
Fix binding issue caused by bad conversion between class and enum + android fix returning wrong dayiteration

ScottBTR committed:
Changes List below:

- RecurrenceRuleReadOnly removed at this stage.
- enums with .None or .NotSet have been altered to no longer include those values and instead values that require a default non-set value have been made nullables
- iOS recurrence rules:
  * Now properly deals with .Last (-1 not 4)
  * Moved Recurrence rule setters into separate function to reduce duplicated code between create and update
  * Saving recurring events now effects all instances of events instead of only the one being edited
- UWP will now automatically add http:// when not included to reduce pain for developers/users as uwp requires a URL instead of string value
- DayIterationOffsetPosition --> WeekOfMonth for readability
  * IterationOffset Enum (should be name changed too) is now set up by platform to allow for easy straight conversions.
- MonthOfTheYear --> MonthOfYear

ScottBTR committed:
Set up for some expected pass test cases, will be refactored at a later date.

ScottBTR committed:
PR feedback alterations.
